### PR TITLE
[stable/stolon] Init script support

### PR DESCRIPTION
**Pull request by mfvitale**
#### What this PR does / why we need it:
The PR will add the support for run script after keeper creation and script after cluster is ready (i.e create user or db). 

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [X ] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [X ] Chart Version bumped
- [ X] Variables are documented in the README.md
- [X ] Title of the PR starts with chart name (e.g. `[stable/chart]`)

@lwolf @rtluckie

### CODE REVIEW

**Comment by demisx** [comment]
Can this be merged? Having init script support will be very useful.

File: stable/stolon/Chart.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v1
 name: stolon
-version: 1.1.2
+version: 1.1.3
```
Comment (Aisuko) [inline stable/stolon/Chart.yaml]: Please bump the minor version to `1.2.0`

File: stable/stolon/templates/configmap.yaml [hunk 2]
```diff
@@ -30,3 +30,6 @@ data:
 
     echo "Node ${NODE_NAME} is ready to shutdown"
 {{- end }}
+{{- with .Values.nodePostStartScript }}
+{{ toYaml . | indent 2 }}
+{{ end }}
```
Comment (Aisuko) [inline stable/stolon/templates/configmap.yaml]: `{{- end}}` can make sense more than `{{ end }}`

File: stable/stolon/templates/hooks/init-db-job.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: init-scripts
+data:
+{{- with .Values.initdbScripts }}
+{{ toYaml . | indent 2 }}
+{{ end }}
```
Comment (Aisuko) [inline stable/stolon/templates/hooks/init-db-job.yaml]: Same as above.

File: stable/stolon/templates/hooks/init-db-job.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      volumes:
+        - name: sql-script
+          configMap:
+            name: init-scripts
+            defaultMode: 420
+      initContainers:
+        - name: wait-for-database
+          image: jwilder/dockerize
+          command: ['dockerize', '-timeout', '30s', '-wait', 'tcp://{{ template "stolon.fullname" . }}-proxy:5432']
+{{ end }}
```
Comment (Aisuko) [inline stable/stolon/templates/hooks/init-db-job.yaml]: Same as above

File: stable/stolon/values.yaml [hunk 5]
```diff
@@ -132,3 +132,21 @@ sentinel:
...
+#   create_script.sh: |
+#      #!/bin/sh
+#      echo "Do something."
+
+## nodePostStart scripts
+## Specify dictionary of scripts to be run at first boot, the entry point script is postStartScript.sh
+## i.e. you can create tablespace directory here.
+##
+# nodePostStartScript:
+#  postStartScript.sh: |-
```
Comment (Aisuko) [inline stable/stolon/values.yaml]: Why there need a `|-`  ?  The scripts don't need to remove the blank in the end right?

**Review by Aisuko — CHANGES_REQUESTED** [review]
Thanks for your PR, there may some advice, please have a look.

File: stable/stolon/templates/configmap.yaml (hunk 2 above)
Comment (mfvitale) [reply stable/stolon/templates/configmap.yaml]: Please can you explain me the meaning of "-".

File: stable/stolon/values.yaml (hunk 5 above)
Comment (mfvitale) [reply stable/stolon/values.yaml]: You are right. Fixed

File: stable/stolon/templates/hooks/init-db-job.yaml (hunk 4 above)
Comment (mfvitale) [reply stable/stolon/templates/hooks/init-db-job.yaml]: Fixed

File: stable/stolon/templates/hooks/init-db-job.yaml (hunk 3 above)
Comment (mfvitale) [reply stable/stolon/templates/hooks/init-db-job.yaml]: fixed

File: stable/stolon/templates/configmap.yaml (hunk 2 above)
Comment (mfvitale) [reply stable/stolon/templates/configmap.yaml]: Fixed

File: stable/stolon/Chart.yaml (hunk 1 above)
Comment (mfvitale) [reply stable/stolon/Chart.yaml]: Fixed

**Comment by mfvitale** [comment]
> Thanks for your PR, there may some advice, please have a look.

Thanks for your advice.

File: stable/stolon/templates/keeper-statefulset.yaml [hunk 6]
```diff
@@ -128,11 +128,23 @@ spec:
           - name: config
             mountPath: /pre-stop-hook.sh
             subPath: pre-stop-hook.sh
+{{ end }}
```
Comment (Aisuko) [inline stable/stolon/templates/keeper-statefulset.yaml]: Replace  `{{ end }}` to `{{- end}}`

File: stable/stolon/templates/keeper-statefulset.yaml [hunk 7]
```diff
@@ -128,11 +128,23 @@ spec:
...
+{{- if .Values.nodePostStartScript }}
+          - name: config
+            mountPath: /postStartScript.sh
+            subPath: postStartScript.sh
+{{ end }}
```
Comment (Aisuko) [inline stable/stolon/templates/keeper-statefulset.yaml]: Replace  `{{ end }}` to `{{- end}}`

File: stable/stolon/templates/keeper-statefulset.yaml [hunk 8]
```diff
@@ -128,11 +128,23 @@ spec:
...
           lifecycle:
+{{- if .Values.keeper.hooks.failKeeper.enabled }}
             preStop:
               exec:
                 command: ["/bin/bash", "-e", "/pre-stop-hook.sh"]
 {{ end }}
```
Comment (Aisuko) [inline stable/stolon/templates/keeper-statefulset.yaml]: Replace  `{{ end }}` to `{{- end}}`

File: stable/stolon/templates/keeper-statefulset.yaml [hunk 9]
```diff
@@ -128,11 +128,23 @@ spec:
...
+{{- if .Values.nodePostStartScript }}
+            postStart:
+              exec:
+                command: ["/bin/bash", "-e", "/postStartScript.sh"]
+{{ end }}
```
Comment (Aisuko) [inline stable/stolon/templates/keeper-statefulset.yaml]: Replace  `{{ end }}` to `{{- end}}`

Comment (Aisuko) [reply stable/stolon/templates/keeper-statefulset.yaml]: Should we reformat line number from 160 to 178 ?

**Review by Aisuko — CHANGES_REQUESTED** [review]

File: stable/stolon/templates/keeper-statefulset.yaml (hunk 9 above)
Comment (mfvitale) [reply stable/stolon/templates/keeper-statefulset.yaml]: > Should we reformat line number from 160 to 178 ?

Seems ok. I haven't touched it

Comment (mfvitale) [reply stable/stolon/templates/keeper-statefulset.yaml]: > Replace `{{ end }}` to `{{- end}}`

Solved

File: stable/stolon/templates/keeper-statefulset.yaml (hunk 8 above)
Comment (mfvitale) [reply stable/stolon/templates/keeper-statefulset.yaml]: Solved

File: stable/stolon/templates/keeper-statefulset.yaml (hunk 7 above)
Comment (mfvitale) [reply stable/stolon/templates/keeper-statefulset.yaml]: Solved

File: stable/stolon/templates/keeper-statefulset.yaml (hunk 6 above)
Comment (mfvitale) [reply stable/stolon/templates/keeper-statefulset.yaml]: Solved

**Comment by Aisuko** [comment]
cc @paulczar Hello Sir, please help us review this.

**Comment by mfvitale** [comment]
@Aisuko @paulczar  any news?

**Comment by Aisuko** [comment]
cc @maorfr, Hello sir, please help us with this.

**Comment by Aisuko** [comment]
Many thanks for you all.

**Comment by dclark** [comment]
nodePostStartScript working well, previously was using a keeper patch to call the init scripts to copy over SSL certificates and set permissions.

**Pull request merged**